src/number_theory/legendre_symbol/mul_character.lean

```diff
@@ -0,0 +1,495 @@
+
+namespace mul_char
+
+universes u v w
+
+variables {R : Type u} [comm_ring R] {R' : Type v} [comm_ring R'] {R'' : Type w} [comm_ring R'']
+
+/-- A multiplicative character is *nontrivial* if it takes a value `≠ 1` on a unit. -/
+def is_nontrivial (χ : mul_char R R') : Prop := ∃ a : Rˣ, χ a ≠ 1
+
+/-- A multiplicative character is nontrivial iff it is not the trivial character. -/
+lemma is_nontrivial_iff (χ : mul_char R R') : χ.is_nontrivial ↔ χ ≠ 1 :=
+begin
+  split,
+  { intros h₁ h₂,
+    obtain ⟨a, ha⟩ := h₁,
+    rw [h₂, one_apply_coe] at ha,
+    exact ha rfl, },
+  { contrapose!,
+    intro h,
+    change ¬ ∃ a : Rˣ, χ a ≠ 1 at h,
+    push_neg at h,
+    ext a,
+    rw [one_apply_coe a, h a], },
+end
+
+/-- A multiplicative character is *quadratic* if it takes only the values `0`, `1`, `-1`. -/
+def is_quadratic (χ : mul_char R R') : Prop := ∀ a, χ a = 0 ∨ χ a = 1 ∨ χ a = -1
+
+/-- We can post-compose a multiplicative character with a ring homomorphism. -/
+def ring_hom_comp (χ : mul_char R R') (f : R' →+* R'') : mul_char R R'' :=
+{ map_nonunit' :=
+  begin
+    intros a ha,
+    simp only [ring_hom.to_monoid_hom_eq_coe, monoid_hom.to_fun_eq_coe, monoid_hom.coe_comp,
+               ring_hom.coe_monoid_hom, coe_coe, function.comp_app, map_nonunit χ ha,
+               ring_hom.map_zero],
+  end,
+  ..f.to_monoid_hom.comp χ.to_monoid_hom}
+
+@[simp]
+lemma ring_hom_comp_apply (χ : mul_char R R') (f : R' →+* R'') (a : R) :
+  χ.ring_hom_comp f a = f (χ a) := rfl
+
+/-- Composition with an injective ring homomorphism preserves nontriviality. -/
+lemma is_nontrivial.comp {χ : mul_char R R'} (hχ : χ.is_nontrivial)
+  {f : R' →+* R''} (hf : function.injective f) :
+   (χ.ring_hom_comp f).is_nontrivial :=
+begin
+  obtain ⟨a, ha⟩ := hχ,
+  use a,
+  rw [ring_hom_comp_apply, ← ring_hom.map_one f],
+  exact λ h, ha (hf h),
+end
+
+/-- Composition with a ring homomorphism preserves the property of being a quadratic character. -/
+lemma is_quadratic.comp {χ : mul_char R R'} (hχ : χ.is_quadratic) (f : R' →+* R'') :
+  (χ.ring_hom_comp f).is_quadratic :=
+begin
+  intro a,
+  simp only [ring_hom_comp_apply],
+  rcases hχ a with (ha | ha | ha); rw [ha],
+  { left,
+    rw [ring_hom.map_zero], },
+  { right, left,
+    rw [ring_hom.map_one], },
+  { right, right,
+    rw [ring_hom.map_neg, ring_hom.map_one], },
+end
+
+/-- The inverse of a quadratic character is itself. →  -/
+lemma quadratic_char_inv {χ : mul_char R R'} (hχ : χ.is_quadratic) : χ⁻¹ = χ :=
+begin
+  ext x,
+  rw [inv_apply_eq_inv],
+  rcases hχ x with h₀ | h₁ | h₂,
+  { rw [h₀, ring.inverse_zero], },
+  { rw [h₁, ring.inverse_one], },
+  { rw [h₂, (by norm_num : (-1 : R') = (-1 : R'ˣ)), ring.inverse_unit (-1 : R'ˣ)],
+    refl, },
+end
+
+/-- The square of a quadratic character is the trivial character. -/
+lemma quad_char_sq_eq_one {χ : mul_char R R'} (hχ : χ.is_quadratic) : χ ^ 2 = 1 :=
+begin
+  rw [pow_two],
+  nth_rewrite 0 ← quadratic_char_inv hχ,
+  rw [mul_left_inv],
+end
```

Comment:
With the `convert` line, we can signal that the proof works because `χ` is its own inverse.
```suggestion
begin
  convert mul_left_inv _,
  rw [pow_two, hχ.inv]
end
```

Reply:
OK. If it weren't necessary to use `nth_rewrite` in my proof, it would have been a one-liner...